### PR TITLE
RFC: groups full sync

### PR DESCRIPTION
**Pull request by sbernhard**

### CODE REVIEW

**Comment by eviltrout** [comment]
Please provide more context for this change. What does it do? Why is it required?

**Comment by sbernhard** [comment]
Currently the discourse saml plugin adds a user to existing groups but it does not remove them automatically. For removing a user from a group you need to add the groups_to_remove attribute on the IDP site. 

With this change, a user is assigned to exactly the groups which he/she is assigned on the IDP site. If the user is removed from a group at IDP, the user will be removed from the group on diacourse, too. Therefore, this feature is full sync of groups.

**Comment by eviltrout** [comment]
I see, that is a worthwhile change indeed. However before accepting it we will need a couple more changes:

1. The documentation should be updated to explain how to use this feature

2. We'll need some tests to confirm the functionality and to prevent future regressions.

**Comment by sbernhard** [comment]
Thanks for you remark. I want to prepare that and I have one question to you:
the group_fullsync can be  feature which can be used together with groups_to_add and groups_to_remove OR it can be a feature exclusive (= without using groups_to_add and groups_to_remove). 

I tend do use the exclusive approach as it is more clear to the user and easier to understand. What do you think?

**Comment by sbernhard** [comment]
Done @eviltrout :-) Love to contribute to discourse!

File: lib/saml_authenticator.rb [hunk 1]
```diff
@@ -196,23 +196,32 @@ def auto_create_account(result)
...
-    total_group_list = (GlobalSetting.try(:saml_sync_groups_list) || "").split('|').map(&:downcase)
+    groups_fullsync = GlobalSetting.try(:saml_groups_fullsync) || false
     group_attribute = GlobalSetting.try(:saml_groups_attribute) || 'memberOf'
     user_group_list = (attributes[group_attribute] || []).map(&:downcase)
-    groups_to_add = user_group_list + attr('groups_to_add').split(",").map(&:downcase)
-    groups_to_remove = attr('groups_to_remove').split(",").map(&:downcase)
 
-    return if user_group_list.blank? && groups_to_add.blank? && groups_to_remove.blank?
+    if groups_fullsync
+      user_has_groups = user.groups.select{|g| g.automatic == false}.map(&:name).map(&:downcase)
```
Comment (eviltrout) [inline lib/saml_authenticator.rb]:
```suggestion
      user_has_groups = user.groups.where(automatic: false).pluck(:name).map(&:downcase)
```

File: lib/saml_authenticator.rb [hunk 2]
```diff
@@ -196,23 +196,32 @@ def auto_create_account(result)
...
+      if user_has_groups != nil
```
Comment (eviltrout) [inline lib/saml_authenticator.rb]:
```suggestion
      if user_has_groups.present?
```

File: README.md [hunk 3]
```diff
@@ -38,6 +38,12 @@ Add the following settings to your `discourse.conf` file:
 
 - `saml_target_url`
 
+### Group sync
+- `DISCOURSE_SAML_SYNC_GROUPS`: Sync groups. Defaults to false.
+- `DISCOURSE_SAML_GROUPS_ATTRIBUTE`: SAML attribute to use for group sync. Defaults to `memberOf`
+- `DISCOURSE_SAML_GROUPS_FULLSYNC`: Should the assigned groups be completely synced including adding AND removing groups based on the IDP? Defaults to false. If set to true, `DISCOURSE_SAML_SYNC_GROUPS_LIST` and SAML attribute `groups_to_add`/`groups_to_remove` are not used.
+- `DISCOURSE_SAML_SYNC_GROUPS_LIST`: Only groups mentioned in this list are synced if they are referenced by the IDP (in `memberOf` SAML attribue).
```
Comment (eviltrout) [inline README.md]:
```suggestion
- `DISCOURSE_SAML_SYNC_GROUPS_LIST`: Groups mentioned in this list are synced if they are referenced by the IDP (in `memberOf` SAML attribue). Any other groups will not be removed/updated.
```

**Comment by sbernhard** [comment]
Thanks for your suggestions. They all look very good - I guess they are working too? Did the tests run?

**Comment by eviltrout** [comment]
I ran the tests myself and it looks great, thanks!

**Comment by sbernhard** [comment]
Great. Thanks for merging all these changes (there is one left :laughing:)
I guess, this is the point, at which all functionalities I need exists at discourse-saml. Hope you have fun with it the same way as I do :laughing:

**Pull request merged**